framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java

```diff
@@ -522,51 +522,80 @@ public JavaExpression visit(NameExpr expr, JavaExpressionContext context) {
-                FieldAccess fieldAccess =
-                        getFieldJavaExpression(fieldElem, receiverExpr, isOriginalReceiver);
-                TypeElement scopeClassElement =
-                        TypesUtils.getTypeElement(fieldAccess.getReceiver().getType());
-                if (!isOriginalReceiver
-                        && !ElementUtils.isStatic(fieldElem)
-                        && ElementUtils.isStatic(scopeClassElement)) {
-                    throw new ParseRuntimeException(
-                            constructJavaExpressionParseError(
-                                    identifier,
-                                    "a non-static field can't be referenced from a static inner class or enum"));
-                }
-                return fieldAccess;
+            if (fieldElem == null) {
+                // field not found.
+                return null;
             }
 
-            return null;
+            // Construct a FieldAccess expression.
+            if (ElementUtils.isStatic(fieldElem)) {
+                Element classElem = fieldElem.getEnclosingElement();
+                JavaExpression staticClassReceiver = new ClassName(ElementUtils.getType(classElem));
+                return new FieldAccess(staticClassReceiver, fieldElem);
+            } else if (receiverExpr instanceof ClassName) {
+                throw new ParseRuntimeException(
+                        constructJavaExpressionParseError(
+                                fieldElem.getSimpleName().toString(),
+                                "a non-static field cannot have a class name as a receiver."));
+            }
+
+            @SuppressWarnings("interning:not.interned") // Checking for exact object
+            boolean fieldDeclaredInReceiverType = enclosingTypeOfField == receiverExpr.getType();
+            // fieldElem is an instance field
+            if (fieldDeclaredInReceiverType) {
+                // It's an instance field declared in the type of receiverExpr.
+                TypeMirror fieldType = ElementUtils.getType(fieldElem);
+                return new FieldAccess(receiverExpr, fieldType, fieldElem);
+            }
+            if (!(receiverExpr instanceof ThisReference)) {
+                throw new ParseRuntimeException(
+                        constructJavaExpressionParseError(
+                                identifier,
+                                "field declared in an outer type cannot be accessed from an inner type"));
+            }
+            TypeElement receiverTypeElement = TypesUtils.getTypeElement(receiverExpr.getType());
+            if (receiverTypeElement == null || ElementUtils.isStatic(receiverTypeElement)) {
+                throw new ParseRuntimeException(
+                        constructJavaExpressionParseError(
+                                identifier,
+                                "a non-static field declared in an outer class cannot be referenced from a static inner class or enum"));
```

Comment:
This message uses "outer class" whereas the one above used "outer type"; make them consistent.